### PR TITLE
ChunkExecutor: Split out ledger update to a separate stage

### PR DESCRIPTION
**Pull request by msmouse**
### Description
1. `execute_chunk()` is now `enqueue_chunk_by_execution()` + `ledger_update()`; `apply_chunk()` is now `enqueue_chunk_by_transaction_outputs()` + `ledger_update()`
2. the sharded state updates and sharded state cache are passed down to the DB
3. `save_transactions()` and `save_transaction_block()` are unified to one.

### Test Plan

tests and 
replay mainnet to varify the chunk replayer https://github.com/aptos-labs/aptos-core/actions/runs/6477553053

### CODE REVIEW

File: execution/executor-types/src/lib.rs [hunk 1]
```diff
@@ -49,24 +49,67 @@ pub mod state_checkpoint_output;
...
         &self,
         txn_list_with_proof: TransactionListWithProof,
         // Target LI that has been verified independently: the proofs are relative to this version.
         verified_target_li: &LedgerInfoWithSignatures,
         epoch_change_li: Option<&LedgerInfoWithSignatures>,
-    ) -> Result<()>;
+    ) -> Result<()> {
+        self.enqueue_chunk_by_execution(txn_list_with_proof, verified_target_li, epoch_change_li)?;
+
+        self.update_ledger()
```
Comment (msmouse) [inline execution/executor-types/src/lib.rs:64]: @JoshLind  the new interfaces look like this

File: execution/executor-types/src/lib.rs [hunk 2]
```diff
@@ -49,24 +49,67 @@ pub mod state_checkpoint_output;
...
         verified_target_li: &LedgerInfoWithSignatures,
         epoch_change_li: Option<&LedgerInfoWithSignatures>,
+    ) -> Result<()> {
+        self.enqueue_chunk_by_transaction_outputs(
+            txn_output_list_with_proof,
+            verified_target_li,
+            epoch_change_li,
+        )?;
+
+        self.update_ledger()
```
Comment (msmouse) [inline execution/executor-types/src/lib.rs:85]: @JoshLind the new interfaces look like this

File: execution/executor-types/src/executed_chunk.rs [hunk 3]
```diff
@@ -4,149 +4,80 @@
...
-                "New validator set of a given epoch LI does not match local computation"
-            );
-            Ok(Some(epoch_change_li.clone()))
-        } else {
-            ensure!(
-                self.next_epoch_state.is_none(),
-                "End of epoch chunk based on local computation but no EoE LedgerInfo provided."
-            );
-            Ok(None)
+        let mut committed_transactions = vec![];
```
Comment (grao1991) [inline execution/executor-types/src/executed_chunk.rs:65]: nit: with_capacity

File: execution/executor-types/src/state_checkpoint_output.rs [hunk 4]
```diff
@@ -129,4 +129,16 @@ impl StateCheckpointOutput {
...
     }
+
+    pub fn ensure_no_discard_or_retry(&self) -> Result<()> {
+        ensure!(
+            self.txns.to_discard.is_empty(),
+            "Syncing discarded transactions"
+        );
+        ensure!(
+            self.txns.to_retry.is_empty(),
+            "Chunk crosses epoch boundary.",
```
Comment (grao1991) [inline execution/executor-types/src/state_checkpoint_output.rs:140]: Should we just say `Syncing retried transactions` here?
(is exceeding block gas limit also a retry?)

File: storage/aptosdb/src/lib.rs [hunk 5]
```diff
@@ -2170,13 +2064,10 @@ impl DbWriter for AptosDB {
                 txns_to_commit,
                 first_version,
                 latest_in_memory_state.current.usage(),
-                Some(sharded_state_cache),
+                sharded_state_cache,
                 self.skip_index_and_usage,
             )?;
 
-            let _timer = OTHER_TIMERS_SECONDS
```
Comment (grao1991) [inline storage/aptosdb/src/lib.rs:2177]: I need this.

File: storage/storage-interface/src/lib.rs [hunk 6]
```diff
@@ -508,7 +508,7 @@ impl MoveStorage for &dyn DbReader {
 #[allow(unused_variables)]
 pub trait DbWriter: Send + Sync {
     /// Get a (stateful) state snapshot receiver.
-    ///
+    //
```
Comment (grao1991) [inline storage/storage-interface/src/lib.rs:511]: ?

File: execution/executor-types/src/executed_chunk.rs [hunk 7]
```diff
@@ -4,149 +4,80 @@
...
-                Ok(())
-            })
+    pub fn combine(&mut self, rhs: Self) {
+        let Self {
+            result_state,
+            ledger_info,
+            next_epoch_state,
+            ledger_update_output,
+        } = rhs;
+        self.result_state = result_state;
```
Comment (grao1991) [inline execution/executor-types/src/executed_chunk.rs:49]: Do you need to combine the HashMap in the result_state?

Comment (msmouse) [reply execution/executor-types/src/executed_chunk.rs:49]: No the "tail" is managed by state_updates_until_latest_checkpoint?

**Review by areshand — APPROVED** [review]

File: execution/executor-types/src/executed_chunk.rs [hunk 8]
```diff
@@ -4,149 +4,94 @@
...
-            transaction_infos.len(),
-        );
+    pub fn combine(&mut self, rhs: Self) {
+        let Self {
+            result_state,
+            ledger_info,
+            next_epoch_state,
+            ledger_update_output,
+        } = rhs;
+
```
Comment (grao1991) [inline execution/executor-types/src/executed_chunk.rs:49]: I feel we need an assert somewhere to ensure the rhs is right after self?

File: execution/executor-types/src/executed_chunk.rs [hunk 9]
```diff
@@ -4,149 +4,94 @@
...
+        let old_result_state = self.result_state.replace_with(result_state);
+        // TODO(aldenhu): This is very unfortunate. Will revisit soon by remodeling the state diff.
+        if self.result_state.base_version > old_result_state.base_version
+            && old_result_state.base_version != old_result_state.current_version
+        {
+            combine_or_add_sharded_state_updates(
```
Comment (grao1991) [inline execution/executor-types/src/executed_chunk.rs:55]: To complicated to understand, I assume it's tested.

File: execution/executor-types/src/executed_chunk.rs [hunk 10]
```diff
@@ -4,149 +4,94 @@
...
 
-    pub fn maybe_select_chunk_ending_ledger_info(
-        &self,
-        verified_target_li: &LedgerInfoWithSignatures,
-        epoch_change_li: Option<&LedgerInfoWithSignatures>,
-    ) -> Result<Option<LedgerInfoWithSignatures>> {
-        let result_accumulator = self.result_view.txn_accumulator();
+    pub fn result_view(&self) -> ExecutedTrees {
+        ExecutedTrees::new(
+            self.result_state.clone(),
```
Comment (grao1991) [inline execution/executor-types/src/executed_chunk.rs:70]: Can we avoid cloning the hashmap inside?

File: execution/executor-types/src/executed_chunk.rs [hunk 11]
```diff
@@ -4,149 +4,94 @@
...
-            Ok(Some(epoch_change_li.clone()))
-        } else {
-            ensure!(
-                self.next_epoch_state.is_none(),
-                "End of epoch chunk based on local computation but no EoE LedgerInfo provided."
-            );
-            Ok(None)
+        let mut committed_transactions =
+            Vec::with_capacity(self.ledger_update_output.to_commit.len());
+        let mut committed_events = Vec::with_capacity(self.ledger_update_output.to_commit.len());
```
Comment (grao1991) [inline execution/executor-types/src/executed_chunk.rs:80]: Is this a fair estimation for the # of events? shall we assume 2x at least?

**Review by grao1991 — APPROVED** [review]

**Pull request merged**